client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java

```diff
@@ -220,11 +276,20 @@ private CloseableHttpAsyncClient createHttpClient() {
                 //default settings for connection pooling may be too constraining
                 .setMaxConnPerRoute(DEFAULT_MAX_CONN_PER_ROUTE).setMaxConnTotal(DEFAULT_MAX_CONN_TOTAL)
                 .setSSLContext(SSLContext.getDefault())
+                .setUserAgent(USER_AGENT_HEADER_VALUE)
                 .setTargetAuthenticationStrategy(new PersistentCredentialsAuthenticationStrategy());
             if (httpClientConfigCallback != null) {
                 httpClientBuilder = httpClientConfigCallback.customizeHttpClient(httpClientBuilder);
             }
 
+            // Always add metadata header last so that it's not overwritten
+            httpClientBuilder.addInterceptorLast((HttpRequest request, HttpContext context) -> {
+                if (metadataHeaderEnabled) {
+                    request.setHeader(METADATA_HEADER, METADATA_HEADER_VALUE);
```

Comment:
While this is better, I still think we should not be silently allowing leniency in the computation. If the header is enabled, we should have a valid value, or throw an error. My original suggestion was to lazily set the cached header value (it could and should be outside this lambda, since as you say, the enabled flag is per client, we don't need to check it on every request). This way the method computing the header value does not need all of the try/catch for failures (with one exception: we still need to handle the various possible runtimes, but we should narrow those to ClassNotFoundException).

Reply:
The header was already computed outside the lambda. Moving it to a static initializer allows computing it only once instead of at every construction of a client.

Regarding exceptions and leniency, this header is enabled by default and should not be something users should be concerned with, except if they want to opt out telemetry by calling `RestClientBuilder.setMetadataHeaderEnabled(false)`. That's all. In particular we should not require users to change their build system if it strips out the manifests we read the version from.

No error related to this header should ever bubble up to the user, and we can accept it to contain partial information: knowing that an ES Java client is making a request even if we could not find its version is already interesting information as it contributes to aggregated statistics about the Java client. Since we know it can happen, the aggregation logic will take care of it, and so in this context partial information is still valid information.

And we can then work to understand why the version was not found and fix/update the discovery code.

Regarding runtimes, we also have to be robust against errors that go beyond the existence of the class. For example an application may have a language runtime in a older version that does have the class we're checking, but not the field/method we read the version from. We should not fail in that case.

We should actually return partial information in that case too (e.g. "I found Groovy, but couldn't find its version") which this PR currently does not do. I have updated my [playground for runtime sniffing](https://github.com/swallez/jvm-languages-versions/blob/2bc92312a84f608b9e87328e6a9262edda37adb3/src/main/java/RuntimeInfo.java) for this.

WDYT? Does this make sense? Did I miss something?

Reply:
Thanks for additional explanation. I understand the desire to not cause an error for a user, since this is something internal. It's understandable for the runtime. But for the Elasticsearch version or apache version, which make available through the build of our client, it could mask a bug on our end. Of course so could the runtimes; we could have a typo in the class or method names we introspect.

The thing that gives me pause about all this code is there are not really tests for the actual data, only for "was anything written in the response?"  Without tests for each of these runtimes, we can't be sure the code actually works. And for Elasticsearch version, using a fatjar seems like something that may be common, so I think we should expect any users building such a jar to include the appropriate manifest value. However, it is certainly something we can add as a requirement later, since something is better than nothing right now.

Reply:
> Without tests for each of these runtimes, we can't be sure the code actually works.

That's a fair point, and is why I have this "[playground project](https://github.com/swallez/jvm-languages-versions)" to test various runtimes. I see two possible approaches here:
- add all those runtimes as _test_ dependencies of the rest client.
- have a separate project like the one above where we can test more language runtime variants, and have the rest client's `RuntimeInfo` code be a direct copy of the one tested in that separate project.

First approach keeps the ES repo self-contained, but adds some dependencies that, even if limited to tests, add some weight to the build. It also doesn't allow testing different versions of a given language runtime unless we create additional Gradle subprojects.

Second approach allows having a more complex build with a main project containing the `RuntimeInfo` code, and several subprojects for each runtime version we want to test (e.g. scala-2.12, scala-2.13, scala-3). This allows testing `RuntimeInfo` with many combinations.

The drawback of this second approach is the manual operation to copy updated `RuntimeInfo` from that project to the rest client's code base, unless we can somehow automate it.

> it is certainly something we can add as a requirement later, since something is better than nothing right now.

Exactly. This is why I initially hard-coded the version as a string, but the additional maintenance burden that you pointed out led me to use what is available now and consider improving things later, e.g. by having a class generated at build time.

We should note that httpclient uses classpath resource lookup to find its version (e.g. `/org/apache/http/nio/version.properties`). Although not as strong as a hardcoded string, it should survive in a fat jar, contrarily to manifest attributes. So maybe an approach we should consider for ES itself (and the client).